components/raftstore-v2/src/raft/storage.rs

```diff
@@ -240,6 +258,172 @@ impl<ER: RaftEngine> Storage<ER> {
     pub fn set_ever_persisted(&mut self) {
         self.ever_persisted = true;
     }
+
+    pub fn take_gen_snap_task(&mut self) -> Option<GenSnapTask> {
+        self.gen_snap_task.get_mut().take()
+    }
+
+    fn get_tablet_index(&self) -> u64 {
+        match self.region_state.get_state() {
+            PeerState::Tombstone | PeerState::Applying => 0,
+            _ => self.region_state.get_tablet_index(),
+        }
+    }
+
+    /// Cancel generating snapshot.
+    pub fn cancel_generating_snap(&mut self, compact_to: Option<u64>) {
+        let snap_state = self.snap_state.borrow();
+        if let SnapState::Generating {
+            ref canceled,
+            ref index,
```

Comment:
How about redefine the struct for v2? For example, index doesn't need to be atomic.

Reply:
added apply task relative log, the index will be used to store the `last_applied_index`, then storage can use it to judge whether if the snapshot task should be canceled after compact.